### PR TITLE
Add route stat test.

### PR DESCRIPTION
**Pull request by anmic**

### CODE REVIEW

File: pybrake/routes.py [hunk 1]
```diff
@@ -43,28 +44,33 @@ def __init__(self, project_id, project_key, host):
...
-      else:
-        self._stats[statKey] = RouteStat(method, route, status_code, time, ms)
-
-  def _flush(self):
+  def flush(self):
     stats = dict()
     with self._lock:
       stats = self._stats
       self._stats = None
+      self._thread.cancel()
```
Comment (vmihailenco) [inline pybrake/routes.py]: Why this is needed?

File: pybrake/routes.py [hunk 2]
```diff
@@ -43,28 +44,33 @@ def __init__(self, project_id, project_key, host):
...
     if not stats:
       raise ValueError("Stats is empty")
 
     stats_json = json.dumps({"routes": [stat.__dict__ for _, stat in stats.items()]})
     requests.post(self._api_url, data=stats_json, headers=self._airbrake_headers)
+
+  def inc_request(self, method='', route='', status_code=0, time=None, ms=0):
+    self._init()
+
+    tm = datetime.utcnow().replace(second=0, microsecond=0).strftime('%Y-%m-%dT%H:%M:%SZ')
```
Comment (vmihailenco) [inline pybrake/routes.py]: What is this?

File: pybrake/routes.py [hunk 3]
```diff
@@ -43,28 +44,33 @@ def __init__(self, project_id, project_key, host):
...
+    statKey = route_stat_key(method, route, status_code, tm)
+    with self._lock:
+      if statKey in self._stats:
+        stat = self._stats.get(statKey)
+        stat.add(ms)
+      else:
+        routeStat = RouteStat(method, route, status_code, time)
```
Comment (vmihailenco) [inline pybrake/routes.py]: 3 lines above this is called `stat`, not `routeStat`. Unify names and move `Stat.add(ms)` outside if/else

File: pybrake/routes.py (hunk 1 above)
Comment (anmic) [reply pybrake/routes.py]: It is need in case execution flush function manually instead timer's function execution.

File: .travis.yml [hunk 4]
```diff
@@ -10,6 +10,7 @@ python:
 install:
   - pip install -r test-requirements.txt
   - pip install requests
+  - pip install requests_mock
```
Comment (vmihailenco) [inline .travis.yml]: move requests and requests_mock to test-requirements.txt

File: pybrake/routes.py [hunk 5]
```diff
@@ -37,26 +34,15 @@ def __init__(self, project_id, project_key, host):
 
     self._thread = None
     self._lock = Lock()
-    self._flush_period = 15.0
+    self.flush_period = 15.0
```
Comment (vmihailenco) [inline pybrake/routes.py]: revert

File: pybrake/routes.py [hunk 6]
```diff
@@ -68,3 +54,20 @@ def _flush(self):
...
+    self._init()
+
+    statKey = route_stat_key(method, route, status_code, time)
+    with self._lock:
+      if statKey in self._stats:
+        stat = self._stats.get(statKey)
+        stat.add(ms)
+      else:
+        stat = RouteStat(method, route, status_code, time)
+        stat.add(ms)
```
Comment (vmihailenco) [inline pybrake/routes.py]: I already said to move stat.add out

File: pybrake/test_routes.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+from datetime import datetime
+from time import sleep
+import json
+import requests_mock
+import pybrake
+
+def test_route_stat():
+  tm = datetime(2000, 1, 1, 0, 0)
+
+  def _request_callback(request, _):
```
Comment (vmihailenco) [inline pybrake/test_routes.py]: remove `, _`

File: pybrake/test_routes.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    routeStat = json.loads(request.body)
+    assert routeStat == {
+      'routes' : [{
+        'method': 'GET',
+        'route': 'test',
+        'statusCode': 200,
+        'count': 2,
+        'sum': 246,
+        'sumsq': 30258,
+        'time': tm.strftime('%Y-%m-%dT%H:%M:%SZ'),
```
Comment (vmihailenco) [inline pybrake/test_routes.py]: replace `tm.strftime('%Y-%m-%dT%H:%M:%SZ')` with real time

File: pybrake/test_routes.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+      }]
+    }
+
+  host = 'http://test.com'
+  route_stats = pybrake.routes.RouteStats(1, "token1", host)
```
Comment (vmihailenco) [inline pybrake/test_routes.py]: use kwargs

File: pybrake/test_routes.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+  route_stats.flush_period = 1.0
+  with requests_mock.mock() as m:
+    url = "http://test.com/api/v4/projects/1/routes-stats"
+    m.post(url, json=_request_callback)
+
+    route_stats.inc_request("GET", "test", 200, tm, 123)
+    route_stats.inc_request("GET", "test", 200, tm, 123)
+
+    sleep(2)
```
Comment (vmihailenco) [inline pybrake/test_routes.py]: use _thread.join instead if sleeping

File: pybrake/routes.py [hunk 11]
```diff
@@ -35,27 +32,19 @@ def __init__(self, project_id, project_key, host):
     }
     self._api_url = "{}/api/v4/projects/{}/routes-stats".format(host, project_id)
 
-    self._thread = None
+    self.thread = None
```
Comment (vmihailenco) [inline pybrake/routes.py]: revert

File: pybrake/routes.py [hunk 12]
```diff
@@ -35,27 +32,19 @@ def __init__(self, project_id, project_key, host):
...
-    self._init()
-
-    statKey = route_stat_key(method, route, status_code, time)
-    with self._lock:
-      if statKey in self._stats:
-        stat = self._stats.get(statKey)
-        stat.add(ms)
-      else:
-        self._stats[statKey] = RouteStat(method, route, status_code, time, ms)
+  def set_flush_period(self, flush_period):
```
Comment (vmihailenco) [inline pybrake/routes.py]: remove
there is nothing that stops you setting obj._flush = 1231

File: pybrake/routes.py [hunk 13]
```diff
@@ -35,27 +32,15 @@ def __init__(self, project_id, project_key, host):
...
-    self._init()
-
-    statKey = route_stat_key(method, route, status_code, time)
-    with self._lock:
-      if statKey in self._stats:
-        stat = self._stats.get(statKey)
-        stat.add(ms)
-      else:
-        self._stats[statKey] = RouteStat(method, route, status_code, time, ms)
+      self.thread = Timer(self._flush_period, self._flush)
```
Comment (vmihailenco) [inline pybrake/routes.py]: self._thread

File: pybrake/test_routes.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+import pybrake
+
+def test_route_stat():
+  def _request_callback(request, context):
+    return request.body
+
+  route_stats = pybrake.routes.RouteStats(
+    project_id=1, project_key="token1", host='http://test.com')
+  route_stats._flush_period = 1
+  with requests_mock.mock() as m:
```
Comment (vmihailenco) [inline pybrake/test_routes.py]: split into blocks

File: pybrake/routes.py [hunk 15]
```diff
@@ -9,23 +9,20 @@ class RouteStat():
   def __dict__(self):
     return {s: getattr(self, s) for s in self.__slots__ if hasattr(self, s)}
 
-  def __init__(self, method, route, status_code, time, ms):
+  def __init__(self, method, route, status_code, time):
```
Comment (vmihailenco) [inline pybrake/routes.py]: use kwargs

File: pybrake/routes.py [hunk 16]
```diff
@@ -9,23 +9,20 @@ class RouteStat():
...
   def add(self, ms):
     self.count += 1
     self.sum += ms
     self.sumsq += ms * ms
 
-def route_stat_key(method, route, status_code, time):
-  return "{}:{}:{}:{}".format(method, route, status_code, time)
-
 class RouteStats():
   def __init__(self, project_id, project_key, host):
```
Comment (vmihailenco) [inline pybrake/routes.py]: use kwargs

File: pybrake/routes.py [hunk 17]
```diff
@@ -68,3 +53,20 @@ def _flush(self):
...
+    with self._lock:
+      if statKey in self._stats:
+        stat = self._stats.get(statKey)
+      else:
+        stat = RouteStat(method, route, status_code, time)
+        self._stats[statKey] = stat
+
+      stat.add(ms)
+
+def route_stat_key(method, route, status_code, time):
```
Comment (vmihailenco) [inline pybrake/routes.py]: use kwargs

File: pybrake/test_routes.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+  route_stats = pybrake.routes.RouteStats(
+    project_id=1, project_key="token1", host='http://test.com')
+  route_stats._flush_period = 1
+
+  with requests_mock.mock() as m:
+    url = "http://test.com/api/v4/projects/1/routes-stats"
+    m.post(url, json=_request_callback)
+
+    tm = datetime(2000, 1, 1, 0, 0)
+    route_stats.inc_request("GET", "test", 200, tm, 123)
```
Comment (vmihailenco) [inline pybrake/test_routes.py]: I've asked to add 2nd route

File: pybrake/test_routes.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
+from datetime import datetime
+import json
+import requests_mock
+import pybrake
```
Comment (vmihailenco) [inline pybrake/test_routes.py]: replace with `import RouteStats from .routes`

**Pull request merged**